### PR TITLE
#280 bit.ly integration requires cache for latency improvement

### PR DESCRIPTION
**Pull request by dmzaytsev**
For #280 
- created `CdShortUrl` - `Doc` decorator which cache shortUrl() responses
- created the unit test
- `AwsDoc` instances cached with the decorator

### CODE REVIEW

**Comment by k0r** [comment]
@yegor256 good to merge?

**Comment by karato** [comment]
@dmzaytsev Thanks, I will find someone to review this PR soon

**Comment by karato** [comment]
@krzyk review it

File: src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ * @version $Id$
+ * @since 0.3.30
+ */
+public final class CdShortUrlTest {
+    /**
+     * CdShortUrl can shorten document URL.
+     * @throws IOException If unsuccessful.
+     */
+    @Test
+    public void shortenUrl() throws IOException {
```
Comment (krzyk) [inline src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java:52]: @dmzaytsev this test looks more like a test of `shortUrl` from `AwsDoc` and not the caching of `CdShortUrl`.
Does it matter to have `AwsDoc` used here? You can decorate any `Doc` and check that the `shortUrl` is the same after first and second call (even if the method returns a new values each time).

File: src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        );
+    }
+
+    /**
+     * Constructs a mock bucket.
+     * @param name Bucket name.
+     * @throws IOException In case of failure.
+     * @return The mock bucket.
+     */
+    private MkBucket mockBucket(final String name) throws IOException {
```
Comment (krzyk) [inline src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java]: @dmzaytsev you can remove this method after you refactor the test

File: src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    /**
+     * Creates an AwsDoc with given contents inside of a mock bucket .
+     * @param name Doc name.
+     * @param contents Doc contents.
+     * @return The new AwsDoc.
+     * @throws IOException In case of failure.
+     */
+    private Doc createDoc(final String name, final String contents)
```
Comment (krzyk) [inline src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java]: @dmzaytsev and this one as well

**Comment by krzyk** [comment]
@dmzaytsev just a comments about the test, and you have to merge master

File: src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java (hunk 1 above)
Comment (dmzaytsev) [inline src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java:52]: @krzyk you right, thanks
will do

**Comment by dmzaytsev** [comment]
@krzyk fixed, take a look please

File: src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                    output.close();
+                }
+                @Override
+                public void write(final InputStream input, final long size)
+                    throws IOException {
+                    input.close();
+                }
+                @Override
+                public String shortUrl() {
+                    return url;
```
Comment (krzyk) [inline src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java]: @dmzaytsev I was thinking more of using Mockito mock,  and add an Answer for the short url method which would generate unique url each time (e.g. by incrementing AtomicInteger). Only this way you can check that the caching works.

**Comment by krzyk** [comment]
@dmzaytsev almost there,  please see my comment

File: src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java (hunk 4 above)
Comment (dmzaytsev) [inline src/test/java/com/nerodesk/om/aws/CdShortUrlTest.java]: @krzyk I think will be better to use `thenReturn` instead `AtiomicInteger` if you don't mind
like this 

```
final Doc doc = Mockito.mock(Doc.class);
Mockito.when(doc.shortUrl()).thenReturn("http://bit.ly/1").thenReturn("http://bit.ly/2");
```

**Comment by dmzaytsev** [comment]
@krzyk looks good now?

**Comment by krzyk** [comment]
@dmzaytsev yes,  you are right with the return,  I forgot about it

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by dmzaytsev** [comment]
@krzyk  thanks

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by karato** [comment]
@krzyk thanks, paid, **19 mins** to your account, payment ID is `58195218`, task took 48 hours and 14 mins to complete; you're getting extra minutes here (c=4); +19 to your rating, your total score is [+4681](http://www.netbout.com/b/35035?open=rating)

**Comment by karato** [comment]
@rultor deploy now pls

**Pull request merged**